### PR TITLE
Bugfix #649: Ensure one instance of patcher is running

### PR DESCRIPTION
**Pull request by sp0lsh**

### CODE REVIEW

File: Assets/PatchKit Patcher/Scripts/InputArgumentsPatcherDataReader.cs [hunk 1]
```diff
@@ -63,9 +63,15 @@ public PatcherData Read()
             {
                 throw new ApplicationException("Unable to parse app data path from command line.");
             }
-
             data.AppDataPath = MakeAppDataPathAbsolute(relativeAppDataPath);
 
+            string lockFilePath;
+            if (!TryReadArgument("--lockfile", out lockFilePath))
+            {
+                throw new ApplicationException("Unable to parse lock file path from command line.");
```
Comment (genail) [inline Assets/PatchKit Patcher/Scripts/InputArgumentsPatcherDataReader.cs]: --lockfile is something optional for now. We shouldn't stop the patcher if it's not passed in. Please also log in that it's being used. Also log a warning if it cannot be found.

File: Assets/PatchKit Patcher/Scripts/Patcher.cs [hunk 2]
```diff
@@ -74,6 +76,8 @@ public static Patcher Instance
 
         private bool _hasAutomaticallyStartedApp;
 
+        private FileStream _LockFileStream;
```
Comment (genail) [inline Assets/PatchKit Patcher/Scripts/Patcher.cs]: Change _LockFileStream to _lockFileStream

File: Assets/PatchKit Patcher/Scripts/Patcher.cs [hunk 3]
```diff
@@ -372,6 +395,11 @@ private void ThreadExecution(CancellationToken cancellationToken)
             catch (ThreadAbortException)
             {
                 DebugLogger.Log("Patcher thread finished: thread has been aborted.");
+            }            
+            catch (MultipleInstancesException exception)
+            {
+                DebugLogger.LogWarning(exception.Message);
```
Comment (genail) [inline Assets/PatchKit Patcher/Scripts/Patcher.cs]: I think of using LogException here.
@witcher112 What do you think?

File: Assets/PatchKit Patcher/Scripts/Patcher.cs [hunk 4]
```diff
@@ -773,5 +826,5 @@ protected virtual void OnUpdateAppStatusChanged(OverallStatus obj)
                 if (UpdateAppStatusChanged != null) UpdateAppStatusChanged(obj);
             });
         }
-    }
+    }
```
Comment (genail) [inline Assets/PatchKit Patcher/Scripts/Patcher.cs]: Trailing spaces

**Comment by sp0lsh** [comment]
applied requested changes

File: Assets/PatchKit Patcher/Scripts/Patcher.cs [hunk 5]
```diff
@@ -407,7 +435,8 @@ private void ThreadLoadPatcherData()
                         AppDataPath =
                             Application.dataPath.Replace("/Assets",
                                 string.Format("/Temp/PatcherApp{0}", EditorAppSecret)),
-                        OverrideLatestVersionId = EditorOverrideLatestVersionId
+                        OverrideLatestVersionId = EditorOverrideLatestVersionId,
+                        LockFilePath = "patcher.lock"
```
Comment (witcher112) [inline Assets/PatchKit Patcher/Scripts/Patcher.cs]: I think that using lock file in editor is not necessary.

File: Assets/PatchKit Patcher/Scripts/MultipleInstanceException.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using System.Runtime.Serialization;
+
+namespace PatchKit.Unity.Patcher
+{
+    [Serializable]
+    internal class MultipleInstancesException: Exception
```
Comment (witcher112) [inline Assets/PatchKit Patcher/Scripts/MultipleInstanceException.cs]: I see multiple class declarations in MultipleInstanceException and MultipleInstance**s**Exception.

File: Assets/PatchKit Patcher/Scripts/Patcher.cs [hunk 7]
```diff
@@ -438,6 +468,29 @@ private void ThreadLoadPatcherData()
             }
         }
 
+        private void EnsureSingleInstance()
+        {
+            string lockFilePath = Data.Value.LockFilePath;
+            DebugLogger.LogFormat("Opening lock file: {0}", lockFilePath);
+
+            if (File.Exists(lockFilePath))
```
Comment (witcher112) [inline Assets/PatchKit Patcher/Scripts/Patcher.cs]: Maybe check if lockFilePath is null or empty and ignore using it in this case.

File: Assets/PatchKit Patcher/Scripts/Patcher.cs [hunk 8]
```diff
@@ -438,6 +468,29 @@ private void ThreadLoadPatcherData()
...
+                    DebugLogger.Log("Lock file open success");
+                }
+                catch
+                {
+                    throw new MultipleInstancesException("Another instance of Patcher spotted");
+                }
+            }
+            else
+            {
+                DebugLogger.LogException(new Exception("LockFile is missing"));
```
Comment (witcher112) [inline Assets/PatchKit Patcher/Scripts/Patcher.cs]: I think warning is more sufficient here.

File: Assets/PatchKit Patcher/Scripts/Patcher.cs (hunk 5 above)
Comment (sp0lsh) [reply Assets/PatchKit Patcher/Scripts/Patcher.cs]: Causes any erros? Allows you to test whole program and avoids edge cases (KISS)

Comment (witcher112) [reply Assets/PatchKit Patcher/Scripts/Patcher.cs]: I see potential problem - while recompiling scripts, Unity stops executing scripts in not defined way. We cannot be sure that file handle will be released which will cause an error while entering play mode second time.

Comment (witcher112) [reply Assets/PatchKit Patcher/Scripts/Patcher.cs]: @sp0lsh bump

Comment (sp0lsh) [reply Assets/PatchKit Patcher/Scripts/Patcher.cs]: @witcher112 pong
More explicit suggestion that lock file is disabled for editor would be better.
